### DISCUSSION TITLE
Что считать изменением состояния?

### DISCUSSION THREAD

**Discussion by nergal-perm** [nergal-perm/tg-core#35] [Закрытые обсуждения] [answered]
Очевидный ответ - любые изменения на доске / столе, а также в руках игроков.
Менее очевидный - изменение текущей фазы игры, т.е. текущий ход. Над этим нужно подумать, мне пока непонятно, есть ли в таком представлении структуры игры (как элемента состояния) какие-то плюсы.

**Comment by citerm** [comment]
Мы изначально хотели развивать парадигму использования способностей, потом мы почитали статей и решили попробовать отказаться от раннего выявления абстракций и начать со сценариев, но теперь мы точечно стараемся уложить общие понятие в привычные абстракции.
Мне начинает казаться, что нам нужно все таки вернуться к изначальному плану и раз уж ты решил приступить к выделению требований для движка, то жить в архитектуре способностей и рассматривать другие общие понятия в рамках этой парадигмы.

Что касается вопроса "изменения состояния", мне кажется что изменение состояния это изменение любого элемента игры, которое влечет за собой изменение категории игрока. Тогда мы всегда сможем отследить, что привело к изменению категории клиента (ов), и какие следующие действия доступны для новой категории.

Пример (Быки и Коровы)
функция - сделать попытку
Игрок "отвечающий - делающий ход" может сделать попытку (ввести число)
Игрок "отвечающий - ожидающий" НЕ может сделать попытку (ввести число)
Игрок "проверяющий - ожидающий" НЕ может сделать попытку (ввести число)
Игрок "проверяющий - делающий ход" НЕ может сделать попытку (ввести число)
функция - проверить попытку
Игрок "отвечающий - делающий ход" НЕ может проверить попытку
Игрок "отвечающий - ожидающий" НЕ может проверить попытку
Игрок "проверяющий - ожидающий" НЕ может проверить попытку
Игрок "проверяющий - делающий ход" может проверить попытку

**Reply by nergal-perm** [reply]
Я пошел в сторону определения базовых понятий, максимально общих для всех игр, потому что понял, что у движка есть еще один пользователь - Автор игры. Ему нужно каким-то образом описать правила и механику игры перед тем, как движок позволит Игрокам играть в нее. При этом описания правил и механики должны основываться на общих понятиях, чтобы движок знал, как работать со всеми играми, а не был заточен под какую-то конкретную. 

Хотя ты прав, это в любом случае отступление от первоначального плана :(

**Reply by nergal-perm** [reply]
А если по теме, то _Категория игрока_ - это всегда производная от состояния. При этом какие-то изменения состояния могут не изменить _Категорию_. 

Вообще, наверное, пока не стоит пытаться определить какие-то общие _Категории_ для игроков, лучше пойти по пути описания правил определения доступности действий через состояние игры.

**Reply by citerm** [reply]
Думаю результатом этой дискуссии должен быть примерно следующий ответ:

- Изменение состояния является результатом выполнения определенного действия. 
- Изменение состояния игры это совокупность воздействий на сущности, участвующие в действии. 
- Определенное действие при одинаковых предусловиях приводит к одному и тому же изменению состояния. 
- Изменение состояния участвует в последующем расчете способностей.

**Comment by nergal-perm** [comment, marked as answer]
Закрываю обсуждение, продолжать будем в рамках #60 и [ADR-0002](https://github.com/nergal-perm/tg-core/wiki/0002-persist-state-changes-as-events)
